In [1]:
# EXECUTE FIRST

# computational imports
import numpy as np
import pandas as pd
from ast import literal_eval
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel
import nltk
from nltk.tokenize import sent_tokenize
from nltk import word_tokenize    
nltk.download('averaged_perceptron_tagger')
from sklearn.feature_extraction import text
from nltk.stem import WordNetLemmatizer 
from nltk.corpus import wordnet as wn
import string

# plotting imports
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style("darkgrid")
from scipy.spatial import distance
# for reading files from urls
import urllib.request
# display imports
from IPython.display import display, IFrame
from IPython.core.display import HTML


[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/user/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


# Lesson 13 - Self-Assessment Solutions

# *Self-Assessment: Modularize Fetching Unique Items*

In [2]:

#this is a test dataframe to use
sa1_df = pd.DataFrame({
        'Food': ['Cake', 'Pie', 'Ice Cream'],
        'Flavors': [['Chocolate','Vanilla', 'Marble'], ['Apple', 'Chocolate', 'Cherry'], ['Vanilla', 'Cherry', 'Mint']]
    })
display(sa1_df)

def getUniqueListFromColumn(df, col, returntype = 'string', sort=True):
      
    #stack everything and get the unique values
    stacked = df.apply(lambda x:pd.Series(x[col], dtype='object'),axis=1).stack().unique()
    
    #if the user wants it sorted, sort it
    if sort:
        stacked = np.sort(stacked)
    
    #if the user wants a string back, join to give a string
    if returntype == 'string':
         stacked = ', '.join(stacked)
    
    return stacked

print(f"The sorted string list is: {getUniqueListFromColumn(sa1_df, 'Flavors', 'string', True)}")
print(f"The unsorted array is: {getUniqueListFromColumn(sa1_df, 'Flavors', 'array', False)}")


Food                       Flavors
0       Cake  [Chocolate, Vanilla, Marble]
1        Pie    [Apple, Chocolate, Cherry]
2  Ice Cream       [Vanilla, Cherry, Mint]

The sorted string list is: Apple, Cherry, Chocolate, Marble, Mint, Vanilla
The unsorted array is: ['Chocolate' 'Vanilla' 'Marble' 'Apple' 'Cherry' 'Mint']


# *Self-Assessment: Load and Display - Solution*

There's nothing too new here. You've done this kind of work before. What's more important here than the code is making sure you take a minute or two to understand the data you're pulling in. What columns do you have available to you? Which columns contain simple values and which columns contain lists. Think about how you could or couldn't use this data to make recommendations.

In [3]:
import pandas as pd
import numpy as np
from ast import literal_eval

ted = pd.read_csv('./data/ted_clean.csv')
#we need ratings to be literally evaluated before using it
ted['ratings'] = ted['ratings'].apply(literal_eval)
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  \
0      Ken Robinson: Do schools kill creativity?            1   
1           Al Gore: Averting the climate crisis            1   
2                  David Pogue: Simplicity sells            1   
3             Majora Carter: Greening the ghetto            1   
4  Hans Rosling: The best stats you've ever seen            1   

        published_date                                            ratings  \
0  2006-06-27 00:11:00  [funny, beautiful, ingenious, courageous, long...   
1  2006-06-27 00:11:00  [funny, courageous, confusing, beautiful, unco...   
2  2006-06-27 00:11:00  [funny, courageous, ingenious, beautiful, unco...   
3  2006-06-27 00:11:00  [courageous, beautiful, confusing, funny, inge...   
4  2006-06-27 20:38:00  [ingenious, funny, beautiful, courageous, long...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  published_year  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110            2006  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520            2006  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292            2006  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550            2006  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869            2006

# *Self-Assessment: Pandas - Solution*
Remember that shape gives you the number of rows first, followed by the number of columns.

In [4]:
ted.shape

(2550, 18)

There are 2550 TED talks in this data frame.

# *Self-Assessment: Prerequisites - Solution*

Remember that when you're calculating the quantile for some piece of data, you'll get different results if you calculate it before or after you do your other subsetting. First, let's calculate the views quantile before we figure the rest of our prerequisites.

In [5]:
#Calculate the number of views for the 10th percentile - calculated from the whole dataframe
m = ted['views'].quantile(0.10)

#Only consider talks of at least 5 minutes
q_talks = ted[(ted['duration'] >= 300)]

#Only consider talks with one speaker
q_talks = q_talks[q_talks['num_speaker']==1]

#Only consider talks in the top 90%
q_talks = q_talks[q_talks['views'] >= m]

#Inspect the number of talks that made the cut
q_talks.shape[0]

2107

Let's compare that with calculating the quantile after we subset.

In [6]:
#Only consider talks of at least 5 minutes
q_talks2 = ted[(ted['duration'] >= 300)]

#Only consider talks with one speaker
q_talks2 = q_talks2[q_talks2['num_speaker']==1]

#Calculate the number of views for the 10th percentile - calculated from the whole dataframe
m2 = q_talks2['views'].quantile(0.10)

#Only consider talks in the top 90%
q_talks2 = q_talks2[q_talks2['views'] >= m2]

#Inspect the number of talks that made the cut
q_talks2.shape[0]

2093

There is no universally "right" answer as to whether you should calculate the quantile before or after you've narrowed the initial dataset. It depends on what you're trying to accomplish. If you want the most viewed talks *that meet your criteria* you'd calculate it after you've subsetted. If you want the most viewed talks *overall* you'd calculate it before you've subsetted.

For our homework, we'll either tell you when to subset a dataframe or ask you to make the decision and give a justification for your decision.

# *Self-Assessment: Compute a Metric, Sort and Print - Solution*

Note that here we are computing our metric on our narrowed data set. We could have created the metric on the entire dataset. But, if we know that we're only interested in a portion of the talks, we should narrow our dataset before computing the metric.

In [7]:
#create the metric of the comments to views ratio
q_talks['comments_per_1000views']=1000*q_talks['comments']/q_talks['views']

#Sort talks in descending order of the ratio of views to comments
q_talks = q_talks.sort_values('comments_per_1000views', ascending=False)

#Print the top 10 talks
q_talks[['description', 'main_speaker', 'comments_per_1000views']].head(10)


description  \
803   David Bismark demos a new system for voting th...   
96    Richard Dawkins urges all atheists to openly s...   
694   Filmmaker Sharmeen Obaid-Chinoy takes on a ter...   
954   Janet Echelman found her true voice as an arti...   
840   Lesley Hazleton sat down one day to read the K...   
1787  Our consciousness is a fundamental aspect of o...   
661   Vaccine-autism claims, "Frankenfood" bans, the...   
443   At TED2009, Al Gore presents updated slides fr...   
732   In "THE 99," Naif Al-Mutawa's new generation o...   
1148  What does environmental devastation actually l...   

               main_speaker  comments_per_1000views  
803           David Bismark                1.534355  
96          Richard Dawkins                1.463841  
694   Sharmeen Obaid-Chinoy                1.420683  
954          Janet Echelman                1.359572  
840         Lesley Hazleton                1.285149  
1787         David Chalmers                1.235918  
661         Michael Specter                1.235704  
443                 Al Gore                1.091008  
732          Naif Al-Mutawa                0.981252  
1148             Garth Lenz                0.977574

# *Self-Assessment: Create the Knowledge-Based Recommender - Solution*

We're creating this as a function that takes in the dataframe and the percentile of views that we want to return. 
We'll first generate our list of unique words to present to users. We'll also stringify our list of ratings so we can use str.contains to filter.

In [8]:
def build_chart(gen_ted, percentile=0.1):
   
    #Ask for preferred word rating
    print("Select a descriptive word from the list above for the 'word rating'")
    rating = input()
    
    #Ask for lower limit of film year
    print("Input earliest year published (2006 to 2017)")
    low_year = int(input())
    
    #Ask for upper limit of film year
    print("Input latest year published(2006 to 2017)")
    high_year = int(input())
    
    
    #Define a new talks variable to store the preferred talks. 
    #Copy the contents of gen_ted to talks
    talks = gen_ted.copy()
    
    #Filter based on the condition
    talks = talks[(talks['ratings'].apply(lambda x: rating in x)) & 
                    (talks['published_year'] >= low_year) & 
                    (talks['published_year'] <= high_year)]
    
    #Calculate the number of views for the  percentile 
    m = talks['views'].quantile(percentile)

    #Only consider movies that have higher than m votes. Save this in a new dataframe q_movies (note using .loc here prevents a warning)
    q_talks = talks.copy().loc[talks['views'] >= m]
    
    #create the metric of the comments to views ratio
    q_talks['comments_per_1000views']=1000*q_talks['comments']/q_talks['views']

    #Sort talks in descending order of the ratio of views to comments
    q_talks = q_talks.sort_values('comments_per_1000views', ascending=False)
    
    return q_talks

In [9]:
#First we'll print a list of possible ratings
r = getUniqueListFromColumn(ted, 'ratings', 'string', True)
print(f'Please select a rating from the following: {r}')

#Generate the chart for top talks for these user preferences and display top 5.
#Show the results for the word rating "obnoxious" and published years between 2009 and 2014.
gen_ted_final = build_chart(ted).head(5)

gen_ted_final[['main_speaker','name','published_year','comments_per_1000views']]

Please select a rating from the following: beautiful, confusing, courageous, fascinating, funny, informative, ingenious, inspiring, jaw-dropping, longwinded, obnoxious, ok, persuasive, unconvincing
Select a descriptive word from the list above for the 'word rating'


 confusing

Input earliest year published (2006 to 2017)


 2010

Input latest year published(2006 to 2017)


 2016

main_speaker  \
694   Sharmeen Obaid-Chinoy   
954          Janet Echelman   
840         Lesley Hazleton   
1787         David Chalmers   
661         Michael Specter   

                                                   name  published_year  \
694   Sharmeen Obaid-Chinoy: Inside a school for sui...            2010   
954        Janet Echelman: Taking imagination seriously            2011   
840               Lesley Hazleton: On reading the Koran            2011   
1787  David Chalmers: How do you explain consciousness?            2014   
661       Michael Specter: The danger of science denial            2010   

      comments_per_1000views  
694                 1.420683  
954                 1.359572  
840                 1.285149  
1787                1.235918  
661                 1.235704

# *Self-Assessment: TF-IDF Vectors - Solution*
This is all straight from the book. More information about the TfidfVectorizer is available online here: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

In [10]:
#Define a TF-IDF Vectorizer Object. Remove all english stopwords
tfidf = TfidfVectorizer(stop_words='english', lowercase=True, ngram_range=(2,2))

#Replace NaN with an empty string
ted['description'] = ted['description'].fillna('')

#Construct the required TF-IDF matrix by applying the fit_transform method on the description feature
tfidf_matrix = tfidf.fit_transform(ted['description'])

#Output the shape of tfidf_matrix (rows first, then columns)
tfidf_matrix.shape

(2550, 63416)

In [11]:
#bonus - take a look some of the individual words in the description
feature_names = tfidf.get_feature_names()
feature_names[500:510]

/home/user/.local/lib/python3.8/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


['40 video',
 '40 years',
 '400 metric',
 '400 pounds',
 '400 years',
 '4000 year',
 '404 page',
 '404 pages',
 '413 billion',
 '45 story']

In [12]:
#bonus - this is saying that for the first document, none of the 500th to 510th words shown above show up in that document
tfidf_list = tfidf_matrix.toarray()
tfidf_list[0, 500:510]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

# *Self-Assessment: Create the Content-Based Recommender Based on Dot Product - Solution*
This is also straight from the book. We don't expect you to understand everything to do with linear kernels. But if you're interested, the documentation is here:
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.linear_kernel.html



In [13]:
# Compute the dot product similarity matrix
sim_matrix = linear_kernel(tfidf_matrix, tfidf_matrix)



In [14]:
def content_recommender(df, seed, seedCol, sim_matrix,  topN=2): 
    #get the indices based off the seedCol
    indices = pd.Series(df.index, index=df[seedCol]).drop_duplicates()
    
    # Obtain the index of the item that matches our seed
    idx = indices[seed]
    
    # Get the pairwsie similarity scores of all items and convert to tuples
    sim_scores = list(enumerate(sim_matrix[idx]))
    
    #delete the item that was passed in
    del sim_scores[idx]
    
    # Sort the items based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    
    # Get the scores of the top-n most similar items.
    sim_scores = sim_scores[:topN]
    
    # Get the item indices
    movie_indices = [i[0] for i in sim_scores]
    
    # Return the topN most similar items
    return df.iloc[movie_indices]


In [15]:
#Get recommendations for Tyler Cowen: Be suspicious of simple stories
content_recommender(ted, 'Tyler Cowen: Be suspicious of simple stories', 'name', sim_matrix, 10)

comments                                        description  duration  \
107         87  Marine biologist Tierney Thys asks us to step ...      1001   
1542       308  Building a skyscraper? Forget about steel and ...       742   
2361        31  Stories are necessary, but they're not as magi...       766   
1631       182  There's an irony behind the latest efforts to ...      1091   
1856        28  In this intriguing talk, biologist Ameenah Gur...       852   
968        235  At TED's Full Spectrum Auditions, comedian Jos...       351   
2042       244  At the end of our lives, what do we most wish ...      1147   
1537      1640  It's a standard assumption in the West: As a s...      1237   
1275       254  "Is it okay if I totally trash your office?" I...       892   
2001       165  When writer Roxane Gay dubbed herself a "bad f...       688   

                        event   film_date  languages         main_speaker  \
107                   TED2003  1044144000         22         Tierney Thys   
1542                  TED2013  1361923200         31        Michael Green   
2361            TEDWomen 2016  1477526400         20      Sisonke Msimang   
1631                  TED2013  1363219200         31        Jared Diamond   
1856           TEDGlobal 2014  1414368000         23  Ameenah Gurib-Fakim   
968   Full Spectrum Auditions  1306195200         39       Joshua Walters   
2042                  TED2015  1426809600         31            BJ Miller   
1537           TEDGlobal 2013  1370995200         30           Eric X. Li   
1275           TEDGlobal 2012  1340755200         29            Elyn Saks   
2001            TEDWomen 2015  1432771200         29           Roxane Gay   

                                                   name  num_speaker  \
107           Tierney Thys: Swim with the giant sunfish            1   
1542  Michael Green: Why we should build wooden skys...            1   
2361   Sisonke Msimang: If a story moves you, act on it            1   
1631   Jared Diamond: How societies can grow old better            1   
1856  Ameenah Gurib-Fakim: Humble plants that hide s...            1   
968          Joshua Walters: On being just crazy enough            1   
2042  BJ Miller: What really matters at the end of life            1   
1537        Eric X. Li: A tale of two political systems            1   
1275  Elyn Saks: A tale of mental illness -- from th...            1   
2001          Roxane Gay: Confessions of a bad feminist            1   

           published_date                                            ratings  \
107   2007-05-21 15:59:00  [funny, beautiful, longwinded, unconvincing, i...   
1542  2013-07-09 15:01:10  [inspiring, ingenious, jaw-dropping, informati...   
2361  2017-01-13 16:19:34  [ingenious, inspiring, fascinating, informativ...   
1631  2013-11-25 16:16:55  [informative, longwinded, unconvincing, obnoxi...   
1856  2014-11-04 16:03:07  [longwinded, obnoxious, persuasive, ok, inform...   
968   2011-06-24 15:50:15  [unconvincing, longwinded, obnoxious, inspirin...   
2042  2015-09-10 14:59:01  [beautiful, informative, inspiring, confusing,...   
1537  2013-07-01 15:36:10  [fascinating, informative, courageous, persuas...   
1275  2012-06-29 13:36:36  [courageous, inspiring, jaw-dropping, persuasi...   
2001  2015-06-22 15:22:48  [informative, inspiring, fascinating, beautifu...   

                                          related_talks  \
107   [{'id': 206, 'hero': 'https://pe.tedcdn.com/im...   
1542  [{'id': 1598, 'hero': 'https://pe.tedcdn.com/i...   
2361  [{'id': 462, 'hero': 'https://pe.tedcdn.com/im...   
1631  [{'id': 1313, 'hero': 'https://pe.tedcdn.com/i...   
1856  [{'id': 83, 'hero': 'https://pe.tedcdn.com/ima...   
968   [{'id': 669, 'hero': 'https://pe.tedcdn.com/im...   
2042  [{'id': 1847, 'hero': 'https://pe.tedcdn.com/i...   
1537  [{'id': 1554, 'hero': 'https://pe.tedcdn.com/i...   
1275  [{'id': 189, 'hero': 'https://pe.tedcdn.com/im...   
2001  [{'id': 1089, 'her

## *Self-Assessment: Metadata Recommender*

Reminder: You are using the ratings and the tags. Sanitize both first. Use all the words from each to make the soup.

In [16]:
# Function to sanitize data to prevent ambiguity. It removes spaces and converts to lowercase
def sanitize(x):
    if isinstance(x, list):
        #Strip spaces and convert to lowercase
        return [str.lower(i.replace(" ", "")) for i in x]
    else:
        #Check if director exists. If not, return empty string
        if isinstance(x, str):
            return str.lower(x.replace(" ", ""))
        else:
            return ''

ted = pd.read_csv('./data/ted_clean.csv')

#literal_eval and sanitize both columns
ted['ratings'] = ted['ratings'].apply(literal_eval).apply(sanitize)
ted['tags'] = ted['tags'].apply(literal_eval).apply(sanitize)

#Function that creates a soup out of the desired metadata
def create_soup(x):
    return ' '.join(x['ratings']) + ' ' + ' '.join(x['tags'])

#create a column with the soup in it    
ted['soup'] = ted.apply(create_soup, axis=1)   


print(f'The soup for {ted["title"][0]} is: \n{ted["soup"][0]}')


The soup for Do schools kill creativity? is: 
funny beautiful ingenious courageous longwinded confusing informative fascinating unconvincing persuasive jaw-dropping ok obnoxious inspiring children creativity culture dance education parenting teaching


In [17]:
count = CountVectorizer(stop_words='english', lowercase=True)
count_matrix = count.fit_transform(ted['soup'])

#Compute the cosine similarity score 
cosine_sim = cosine_similarity(count_matrix, count_matrix)

#call our same function, using the same movie. 
content_recommender(ted, 'Humble plants that hide surprising secrets', 'title', cosine_sim, topN=5)

comments                                        description  duration  \
786       954  Plants behave in some oddly intelligent ways: ...       830   
575        56  While living and working as a marine biologist...       998   
584        58  Nick Veasey shows outsized X-ray images that r...       798   
793       266  Did you know you have functioning neurons in y...       914   
805       124  In the quest to map the brain, many scientists...      1054   

              event   film_date  languages      main_speaker  \
786  TEDGlobal 2010  1279152000         30   Stefano Mancuso   
575   TEDIndia 2009  1257379200         25  Charles Anderson   
584  TEDGlobal 2009  1248393600         20       Nick Veasey   
793  TEDGlobal 2010  1279238400         29   Heribert Watzke   
805  TEDGlobal 2010  1279238400         28  Gero Miesenboeck   

                                                  name  num_speaker  \
786   Stefano Mancuso: The roots of plant intelligence            1   
575  Charles Anderson: Dragonflies that fly across ...            1   
584                Nick Veasey: Exposing the invisible            1   
793             Heribert Watzke: The brain in your gut            1   
805         Gero Miesenboeck: Re-engineering the brain            1   

          published_date                                            ratings  \
786  2010-10-11 08:55:00  [persuasive, inspiring, informative, fascinati...   
575  2009-12-17 09:06:00  [courageous, fascinating, informative, ingenio...   
584  2010-01-05 09:13:00  [fascinating, beautiful, funny, jaw-dropping, ...   
793  2010-10-19 08:09:00  [inspiring, informative, fascinating, unconvin...   
805  2010-11-03 22:44:00  [informative, jaw-dropping, fascinating, ingen...   

                                         related_talks    speaker_occupation  \
786  [{'id': 509, 'hero': 'https://pe.tedcdn.com/im...  Plant neurobiologist   
575  [{'id': 416, 'hero': 'https://pe.tedcdn.com/im...      Marine biologist   
584  [{'id': 643, 'hero': 'https://pe.tedcdn.com/im...       X-ray visionary   
793  [{'id': 976, 'hero': 'https://pe.tedcdn.com/im...        Food scientist   
805  [{'id': 967, 'hero': 'https://pe.tedcdn.com/im...        Optogeneticist   

                                                  tags  \
786  [biology, botany, collaboration, plants, science]   
575  [biodiversity, biology, birds, ecology, insect...   
584                [art, nature, photography, science]   
793               [biology, food, science, technology]   
805            [biology, brain, neuroscience, science]   

                                  title  \
786     The roots of plant intelligence   
575  Dragonflies that fly across oceans   
584              Exposing the invisible   
793               The brain in your gut   
805            Re-engineering the brain   

                                                   url    views  \
786  https://www.ted.com/talks/stefano_mancuso_the_...  1075196   
575  https://www.ted.com/talks/charles_anderson_dis...   553253   
584  https://www.ted.com/talks/nick_veasey_exposing...  1780196   
793  https://www.ted.com/talks/heribert_watzke_the_...  1342475   
805       https://www.ted.com/talks/gero_miesenboeck\n   611081   

     published_year                                               soup  
786            2010  persuasive inspiring informative fascinating l...  
575            2009  courageous fascinating informative ingenious b...  
584            2010  fascinating beautiful funny jaw-dropping confu...  
793            2010  inspiring informative fascinating unconvincing...  
805            2010  informative jaw-dropping fascinating ingenious...